packages/filestore-contract/src/partitioned.ts

```diff
@@ -6,23 +6,156 @@ import {
   UnorderedMap,
 } from "near-sdk-js";
 
+interface FileSet {
+  files: string[];
+  storageSize: number;
+}
+
+class Application {
+  currentVersion: number = 0;
+  nextVersion: number | null = null;
+  currentFiles: string[] = [];
+  nextFiles: string[] = [];
+  previousFiles: string[] = [];
+}
+
+const BLANK_APPLICATION_BYTES = 65; // includes key prefix; just add app name length
+
 @NearBindgen({})
 class StatusMessage {
+  applications: UnorderedMap<Application>;
   filemap: UnorderedMap<number>;
   partitions: UnorderedMap<string>;
   constructor() {
+    this.applications = new UnorderedMap("apps");
     this.filemap = new UnorderedMap("pieces");
     this.partitions = new UnorderedMap("parts");
   }
 
+  buildApplicationKey(application: string, account: string, ) {
+    return `${account}/${application}`;
+  }
+
+  buildFileSet(app: Application | null, files: string[]) {
+    return files.reduce((fileSet: FileSet, f: string) => {
+      const file = `v${app?.nextVersion || 1}_${f}`;
+      fileSet.storageSize += file.length;
+      fileSet.files.push(file);
+      return fileSet;
+    }, { files: [], storageSize: 0 });
+  }
+
+  startDeployment({ app, account_id, application, files }: { app: Application, account_id: string, application: string, files: string[] }) {
+    app.nextVersion = app.currentVersion + 1;
+    const { files: nextFiles } = this.buildFileSet(app, files);
+    app.nextFiles = nextFiles;
+    this.applications.set(this.buildApplicationKey(application, account_id), app);
+  }
+
+  completeDeployment({ app, account_id, application }: { app: Application, account_id: string, application: string }) {
+    // @ts-ignore
+    app.previousFiles = (app?.previousFiles || []).concat(app?.currentFiles || [])
+    app.currentFiles = app.nextFiles;
```

Comment:
I guess we should do a check on the `app` argument before calling this function, so that `app` would always be defined as the interface suggests. Otherwise, the next next lines would also require an optional chaining.